File: P0053_Max_Subarray_Sum.cpp
// Problem Link: https://leetcode.com/problems/maximum-subarray/
// Godbolt: https://www.godbolt.org/z/fcqoYhEcK

#include <range/v3/view/partial_sum.hpp>

auto max_sub_array(auto nums) {
    return std::ranges::max(
        ranges::views::partial_sum(nums, 
            [](auto a, auto b) { return std::max(a + b, b); }));
}

File: P0053_Max_Subarray_Sum.kt
// Problem Link: https://leetcode.com/problems/maximum-subarray/
// Kotlin Playground: https://pl.kotl.in/6vqoX0ust

fun maxSubArray(nums: List<Int>): Int {
    return nums.runningReduce{ a, b -> maxOf(a + b, b) }
               .maxOrNull()!!
}
